Update AI Search Flows documentation (#10055)
opensearch-project/documentation-website: _vector-search/ai-search/building-flows.md

---
layout: default
title: Configuring AI search types
parent: Building AI search workflows in OpenSearch Dashboards
grand_parent: AI search
nav_order: 10
---

# Configuring AI search types

> This page provides example configurations for different AI search workflow types. Each example shows how to tailor the setup to a specific use case, such as semantic search or hybrid retrieval. To build a workflow from start to finish, follow the steps in [Building AI search workflows in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/vector-search/ai-search/workflow-builder/), substituting your use case from this page into the appropriate parts of the setup. 

Reviewer: ```suggestion
This page provides example configurations for different AI search workflow types. Each example shows how to tailor the setup to a specific use case, such as semantic search or hybrid retrieval. To build a workflow from start to finish, follow the steps in [Building AI search workflows in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/vector-search/ai-search/workflow-builder/), adding your use case from this page to the appropriate parts of the setup. 
```

> 

Reviewer: Line 11, last sentence: By "this page", do we mean the linked page or this actual page? If the former, it should be "that page".

## Prerequisite: Provision ML resources

> Before you start, select and provision the necessary machine learning (ML) resources depending on your use case. For example, to implement semantic search, you must configure a text embedding model in your OpenSearch cluster. For more information about deploying ML models locally or connecting to externally hosted models, see [Integrating ML models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/integrating-ml-models/).

Reviewer: ```suggestion
Before you start, select and provision the necessary machine learning (ML) resources, depending on your use case. For example, to implement semantic search, you must configure a text embedding model in your OpenSearch cluster. For more information about deploying ML models locally or connecting to externally hosted models, see [Integrating ML models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/integrating-ml-models/).
```

<details markdown="block">
  <summary>
    Table of contents
  </summary>
  {: .text-delta }
1. TOC
{:toc}
</details>

<p id="implementation-examples"></p>

## Semantic search

This example demonstrates how to configure semantic search.

### ML resources

> Create and deploy an [Amazon Titan Text Embeddings model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-bedrock-titan-text-embedding).

Reviewer: ```suggestion
Create and deploy an [Amazon Titan Text Embedding model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-bedrock-titan-text-embedding).
```

```json
{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "<embedding_field_name>": {
        "type": "knn_vector",
        "dimension": "<embedding_size>"
      }
    }
  }
}
```

{% include copy.html %}

### Ingest pipeline

Configure a single ML inference processor. Map your input text to the `inputText` model input field. Optionally, map the output `embedding` to a new document field.

### Search pipeline

Configure a single ML inference search request processor. Map the query field containing the input text to the `inputText` model input field. Optionally, map the output `embedding` to a new field. Override the query to include a `knn` query, for example:

```json
{
    "_source": {
        "excludes": [
            "<embedding_field>"
        ]
    },
    "query": {
        "knn": {
            "<embedding_field>": {
                "vector": ${embedding},
                "k": 10
            }
        }
    }
}
```

{% include copy.html %}

---

## Hybrid search

Hybrid search combines keyword and vector search. This example demonstrates how to configure hybrid search.

### ML resources

> Create and deploy an [Amazon Titan Text Embeddings model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-bedrock-titan-text-embedding).

Reviewer: ```suggestion
Create and deploy an [Amazon Titan Text Embedding model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-bedrock-titan-text-embedding).
```

```json
{
    "_source": {
        "excludes": [
            "<embedding_field>"
        ]
    },
    "query": {
        "hybrid": {
            "queries": [
                {
                    "match": {
                        "<text_field>": {
                            "query": "<text_field_input>"
                        }
                    }
                },
                {
                    "knn": {
                        "<embedding_field>": {
                            "vector": ${embedding},
                            "k": 10
                        }
                    }
                }
            ]
        }
    }
}
```

{% include copy.html %}

**For the normalization processor**, configure weights for each subquery. For more information, see the [hybrid search normalization processor example]({{site.url}}{{site.baseurl}}/search-plugins/hybrid-search/#step-3-configure-a-search-pipeline).

---

## Basic RAG (document summarization)

This example demonstrates how to configure basic retrieval-augmented generation (RAG).

> The following example shows a simplified connector blueprint for the [Claude v1 messages API](https://docs.anthropic.com/en/api/getting-started#examples). While connector blueprints and model interfaces may evolve over time, this example demonstrates how to abstract complex API interactions into a single `prompt` field input.

Reviewer: At the link, only the "Anthropic API" is referenced. Please confirm the API name.

}
```

### ML resources

Create and deploy an [Anthropic Claude 3 Sonnet model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#claude-3-sonnet-hosted-on-amazon-bedrock).

### Search pipeline

Configure an ML inference search response processor using the following steps:

1. Select **Template** as the transformation type for the `prompt` input field.
2. Open the template configuration by selecting **Configure**.
3. Choose a preset template to simplify setup.
4. Create an input variable that extracts the list of reviews (for example, `review`).
5. Inject the variable into the prompt by copying and pasting it into the template.
6. Select **Run preview** to verify that the transformed prompt correctly incorporates sample dynamic data.
7. Select **Save** to apply the changes and exit.

---

## Multimodal search

Multimodal search searches by text and image. This example demonstrates how to configure multimodal search.

### ML resources

> Create and deploy an [Amazon Titan Multimodal Embeddings model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-bedrock-titan-multimodal-embedding).

Reviewer: ```suggestion
Create and deploy an [Amazon Titan Multimodal Embedding model on Amazon Bedrock](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-bedrock-titan-multimodal-embedding).
```

Optionally, map the output `embedding` to a new document field.

### Search pipeline

Configure a single ML inference search request processor. Map the input text field and input image field in the query to the `inputText` and `inputImage` model input fields, respectively. If both text and image inputs are needed, ensure that both are mapped. Alternatively, you can map only one input (either text or image) if a single input is sufficient for embedding generation.

Override the query so that it contains a `knn` query, including the embedding output:

```json
{
    "_source": {
        "excludes": [
            "<embedding_field>"
        ]
    },
    "query": {
        "knn": {
            "<embedding_field>": {
                "vector": ${embedding},
                "k": 10
            }
        }
    }
}
```

{% include copy.html %}

---

## Named entity recognition

This example demonstrates how to configure named entity recognition (NER).

### ML resources

Create and deploy an [Amazon Comprehend Entity Detection model](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-comprehend---entity-detection).

### Ingest pipeline

> Configure a single ML inference processor. Map your input text field to the `text` model input fields. To persist any identified entities with each document, transform the output (an array of entities) and store them in the `entities_found` field. Use the following `output_map` configuration as a reference:

Reviewer: 2nd sentence, last word: Should be it be "field" (sg.)?

> Configure a single ML inference processor. Map your input text field to the `text` model input fields. To persist any identified entities with each document, transform the output (an array of entities) and store them in the `entities_found` field. Use the following `output_map` configuration as a reference:

Reviewer 2: ```suggestion
Configure a single ML inference processor. Map your input text field to the `text` model input field. To persist any identified entities with each document, transform the output (an array of entities) and store them in the `entities_found` field. Use the following `output_map` configuration as a reference:
```

```json
"output_map": [
    {
        "entities_found": "$.response.Entities[*].Type"
    }
],
```

{% include copy.html %}

This configuration maps the extracted entities to the `entities_found` field, ensuring that they are stored alongside each document.

---

## Language detection and classification

The following example demonstrates how to configure language detection and classification.

### ML resources

Create and deploy an [Amazon Comprehend Language Detection model](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#amazon-comprehend---language-detection).

### Ingest pipeline

> Configure a single ML inference processor. Map your input text field to the `text` model input fields. To store the most relevant or most likely language detected for each document, transform the output (an array of languages) and persist it in the `detected_dominant_language` field. Use the following `output_map` configuration as a reference:

Reviewer: 2nd sentence, last word: Should be it be "field" (sg.)?

```json
"input_map": [
   {
      "new_score": "results[0].relevance_score"
   }
],
```

{% include copy.html %}

Use the following rerank processor configuration: Under **target_field**, select the model score field (in this example, `new_score`).

---

## Multimodal search (text or image) with a custom CLIP model

The following example uses a custom CLIP model hosted on Amazon SageMaker. The model dynamically ingests a text or image URL as input and returns a vector embedding.

### ML resources

Create and deploy a [Custom CLIP Multimodal model](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md#custom-clip-multimodal-embedding).

### Index

Ensure that the index settings include `index.knn: true` and that your index contains a `knn_vector` field specified in the mappings, as follows:

```json
{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "<embedding_field_name>": {
        "type": "knn_vector",
        "dimension": "<embedding_size>"
      }
    }
  }
}
```

{% include copy.html %}

### Ingest pipeline

> Configure a single ML inference processor. Map your image field to the `image_url` model input field or your text field to the `text` model input field, depending on what type of data you are ingesting and persisting in your index. For example, if building an application that returns relevant images based on text or image input, you want to persist images and should map the image field to the `image_url` field.

Reviewer: ```suggestion
Configure a single ML inference processor. Map your image field to the `image_url` model input field or your text field to the `text` model input field, depending on what type of data you are ingesting and persisting in your index. For example, if building an application that returns relevant images based on text or image input, you will need to persist images and should map the image field to the `image_url` field.
```
